File: src/modules/lua/test/test010_callbacks.lua
unittest = require "unittest"

function test_key_action_enum_exists()
    assert(mge.key_action ~= nil, "key_action should exist")
    assert(mge.key_action.PRESS ~= nil, "key_action.PRESS should exist")
    assert(mge.key_action.RELEASE ~= nil, "key_action.RELEASE should exist")
    assert(mge.key_action.REPEAT ~= nil, "key_action.REPEAT should exist")
end

function test_add_redraw_listener()
    local MyApp = class(mge.application)
    MyApp.__name__ = "test_callback_app"

    __listener_registered__ = false
    function MyApp:setup()
        self:add_redraw_listener(function(cycle, delta)
            -- not called in this test (no event loop)
        end)
        __listener_registered__ = true
    end

    mge.register_component(mge.application, MyApp)
    local app = mge.create_component(mge.application,
                                     "test_callback_app")
    app:setup()
    assert(__listener_registered__,
        "redraw listener should have been registered")
    app = nil
    collectgarbage("collect")
end

unittest.main()

File: src/modules/python/test/test010_callbacks.py
# mge - Modern Game Engine
# Copyright (c) 2017-2026 by Alexander Schroeder
# All rights reserved.
import unittest
import gc
import mge


class TestCallbacks(unittest.TestCase):

    def tearDown(self):
        gc.collect()

    def test_add_redraw_listener_exists(self):
        app = mge.application()
        self.assertTrue(hasattr(app, 'add_redraw_listener'))

    def test_add_redraw_listener_accepts_callable(self):
        class CallbackApp(mge.application):
            listener_registered = False

            def setup(self):
                def on_redraw(cycle, delta):
                    pass
                self.add_redraw_listener(on_redraw)
                CallbackApp.listener_registered = True

        CallbackApp.__mge_name__ = "test_py_callback_app"
        mge.register_component(mge.application, CallbackApp)
        app = mge.create_component(mge.application, "test_py_callback_app")
        app.setup()
        self.assertTrue(CallbackApp.listener_registered)

    def test_add_redraw_listener_lambda(self):
        class LambdaApp(mge.application):
            registered = False

            def setup(self):
                self.add_redraw_listener(lambda cycle, delta: None)
                LambdaApp.registered = True

        LambdaApp.__mge_name__ = "test_py_lambda_app"
        mge.register_component(mge.application, LambdaApp)
        app = mge.create_component(mge.application, "test_py_lambda_app")
        app.setup()
        self.assertTrue(LambdaApp.registered)
